warehouse/jobs/runner.go

```diff
@@ -0,0 +1,373 @@
+			return nil
+		})
+		g.Wait()
+	}
+
+}
+
+func (asyncWhJob *AsyncJobWhT) cleanUpAsyncTable() error {
+	asyncWhJob.log.Info("WH-Jobs: Cleaning up the zombie asyncjobs")
+	sqlStatement := fmt.Sprintf(`UPDATE %s SET status='%s' WHERE status='%s'`, warehouseutils.WarehouseAsyncJobTable, WhJobWaiting, WhJobExecuting)
+	asyncWhJob.log.Infof("WH-Jobs: resetting up async jobs table query %s", sqlStatement)
+	_, err := asyncWhJob.dbHandle.Query(sqlStatement)
+	return err
+}
+
+/*
+startAsyncJobRunner is the main runner that
+1) Periodically queries the db for any pending async jobs
+2) Groups them together
+3) Publishes them to the pgnotifier
+4) Spawns a subroutine that periodically checks for responses from pgNotifier/slave worker post trackBatch
+*/
+func (asyncWhJob *AsyncJobWhT) startAsyncJobRunner(ctx context.Context) {
+	asyncWhJob.log.Info("WH-Jobs: Starting async job runner")
+	for {
+		asyncWhJob.log.Info("WH-Jobs: Scanning for waiting async job")
+		select {
+		case <-ctx.Done():
+			asyncWhJob.log.Info("WH-Jobs: Stopping AsyncJobRunner")
+			return
+		case <-time.After(10 * time.Second):
+
+		}
+
+		asyncjobpayloads, err := asyncWhJob.getPendingAsyncJobs(ctx)
+		if err != nil {
+			return
+		}
+
+		if len(asyncjobpayloads) > 0 {
+			asyncWhJob.log.Info("WH-Jobs: Got pending wh async jobs")
+			asyncWhJob.log.Infof("WH-Jobs: Number of async wh jobs left = %d\n", len(asyncjobpayloads))
+			notifierClaims, err := getMessagePayloadsFromAsyncJobPayloads(asyncjobpayloads)
+
+			if err != nil {
+				asyncWhJob.log.Errorf("Error converting the asyncJobType to notifier payload %s ", err)
+				asyncWhJob.updateMultipleAsyncJobs(&asyncjobpayloads, WhJobFailed, err.Error())
+				return
+			}
+			messagePayload := pgnotifier.MessagePayload{
+				Jobs:    notifierClaims,
+				JobType: "async_job",
+			}
+			schema := warehouseutils.SchemaT{}
+			ch, err := asyncWhJob.pgnotifier.Publish(messagePayload, &schema, 100)
+			if err != nil {
+				asyncWhJob.updateMultipleAsyncJobs(&asyncjobpayloads, WhJobFailed, err.Error())
+				return
+			}
+			asyncWhJob.updateMultipleAsyncJobs(&asyncjobpayloads, WhJobExecuting, "")
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				//Is this correct or should I be passing ch as a variable to the anonymous go routine instead?
+				responses := <-ch
+				asyncWhJob.log.Info("Response recieved from the pgnotifier track batch")
+				var successfulAsyncJobs []AsyncJobPayloadT
+				for _, resp := range responses {
+					var jobID pgnotifier.AsyncOutput
+					err = json.Unmarshal(resp.Output, &jobID)
+					if err != nil {
+						asyncWhJob.log.Errorf("Unable unmarshal output %s", err)
+						//Will skip to next response, leaving the table row in executing stage.
+						// We need to program retry strategy, timeout strategy for each row in the asyncJobtable
+						continue
+					}
+					ind := getSinglePayloadFromBatchPayloadByTableName(&asyncjobpayloads, jobID.TableName)
+					if ind == -1 {
```

Comment:
Can we set a better name for find. It's not clear what it is and why we have this check of `ind == -1`

Reply:
If that's the case then asyncjobpayloads[ind] will not work.

Reply:
Ideally -1 would not appear right.

Reply:
Yes. -1 would be appear ideally.